pose-detection/src/calculators/relative_velocity_filter.ts

```diff
@@ -0,0 +1,113 @@
+ */
+// ref:
+// https://github.com/google/mediapipe/blob/master/mediapipe/util/filtering/relative_velocity_filter.cc
+export class RelativeVelocityFilter {
+  private lastValue: number = 0;
+  private lastValueScale: number = 1;
+  private lastTimestamp: number = -1;
+
+  private window: WindowElement[] = [];
+  private lowPassFilter: LowPassFilter = new LowPassFilter(1.0);
+
+  /**
+   * Constructor of `RelativeVelocityFilter` class.
+   * @param config
+   *        `windowSize`:  Higher windowSize adds to lag and to stability.
+   *        `velocityScale`: Lower velocityScale adds to lag and to stability.
+   */
+  constructor(private readonly config: VelocityFilterConfig) {}
+
+  /**
+   * Applies filter to the value.
+   * @param value valueToFilter.
+   * @param macroSeconds timestamp associated with the value (for instance,
+   *     timestamp of the frame where you got value from).
+   * @param valueScale value scale (for instance, if your value is a distance
+   *     detected on a frame, it can look same on different devices but have
+   *     quite different absolute values due to different resolution, you
+   *     should come up with an appropriate parameter for your particular use
+   *     case).
+   */
+  apply(value: number, macroSeconds: number, valueScale: number): number {
+    if (value == null) {
+      return value;
+    }
+
+    const $macroSeconds = Math.trunc(macroSeconds);
+    if (this.lastTimestamp >= $macroSeconds) {
+      // Results are unpreditable in this case, so nothing to do but return
+      // same value.
+      return value;
+    }
+
+    let alpha;
+    if (this.lastTimestamp === -1) {
+      alpha = 1;
+    } else {
+      // Implement the DistanceEstimationMode.kLegacyTransition.
+      const distance =
+          value * valueScale - this.lastValue * this.lastValueScale;
+      const duration = $macroSeconds - this.lastTimestamp;
+
+      let cumulativeDistance = distance;
+      let cumulativeDuration = duration;
+
+      // Define max cumulative duration assuming 30 frames per second is a good
+      // frame rate, so assuming 30 values per second or 1 / 30 of a second is
+      // a good duration per window element.
+      const assumedMaxDuration = SECOND_TO_MACRO_SECONDS / 30;
+      const maxCumulativeDuration =
+          (1 + this.window.length) * assumedMaxDuration;
+      for (const el of this.window) {
+        if (cumulativeDuration + el.duration > maxCumulativeDuration) {
+          // This helps in cases when durations are large and outdated
+          // window elements have bad impact on filtering results.
+          break;
+        }
+        cumulativeDistance += el.distance;
+        cumulativeDuration += el.duration;
+      }
+
+      const velocity =
+          cumulativeDistance / (cumulativeDuration * MACRO_SECONDS_TO_SECOND);
+      alpha = 1 - 1 / (1 + this.config.velocityScale * Math.abs(velocity));
+      this.window.push({distance, duration});
```

Comment:
For the loop above it's important where new elements are pushed into the window. You reversed the order in the window, but kept the loop not reversed.